src/Chat/Plugin/Google.php

```diff
@@ -0,0 +1,85 @@
+<?php  declare(strict_types = 1);
+namespace Room11\Jeeves\Chat\Plugin;
+use Amp\Artax\Response;
+use Room11\Jeeves\Chat\Client\ChatClient;
+use Room11\Jeeves\Chat\Command\Command;
+use Room11\Jeeves\Chat\Command\Message;
+class Google implements Plugin {
+    const COMMAND = "google";
+    private $chatClient;
+    private $bitlyAccessToken;
+
+    public function __construct(ChatClient $chatClient, string $bitlyAccessToken) {
+        $this->chatClient = $chatClient;
+        $this->bitlyAccessToken = $bitlyAccessToken;
+    }
+    public function handle(Message $message): \Generator {
+        if (!$this->validMessage($message)) {
+            return;
+        }
+        yield from $this->getResult($message);
+    }
+    private function validMessage(Message $message): bool {
+        return $message instanceof Command
+            && $message->getCommand() === self::COMMAND
+            && $message->getParameters();
+    }
+    private function getResult(Message $message): \Generator {
+        $uri = "https://www.google.com/search?q=" . urlencode(implode(' ', $message->getParameters()));
+        /** @var Response $response */
+        $response = yield from $this->chatClient->request($uri);
+        if ($response->getStatus() !== 200) {
+            yield from $this->chatClient->postMessage(
+                "It was Google's fault, not mine."
+            );
+            return;
+        }
+        $internalErrors = libxml_use_internal_errors(true);
+        $dom = new \DOMDocument();
+        $google = $response->getBody();
+        $dom->loadHTML($google);
+        libxml_use_internal_errors($internalErrors);
+        $xpath = new \DOMXPath($dom);
+        $nodes = $xpath->query("//*[contains(concat(' ', normalize-space(@class), ' '), ' g ')]");
+        if($nodes->length === 0) {
+            yield from $this->chatClient->postMessage(sprintf(
+                ":%s Did you know? That `%s...` dosen't exist in the world! Cuz' GOOGLE can't find it :P",
+                $message->getOrigin(),
+                substr(implode(" ", $message->getParameters()), 0, 60)
+            ));
+            return;
+        }
+        $length = min(3, $nodes->length);
+        $toPostMessage = "";
+        for($i = 0; $i < $length; $i++) {
+            start_of_loop:
```

Comment:
Oh well my bad :( didn't notice those... Gonna fix them in a while..

Gourab Nag
On Apr 23, 2016 10:58 PM, "Niklas Keller" <notifications@github.com> wrote:

> In src/Chat/Plugin/Google.php
> <https://github.com/Room-11/Jeeves/pull/39#discussion_r60831861>:
>
> > +        $dom->loadHTML($google);
> > +        libxml_use_internal_errors($internalErrors);
> > +        $xpath = new \DOMXPath($dom);
> > +        $nodes = $xpath->query("//*[contains(concat(' ', normalize-space(@class), ' '), ' g ')]");
> > +        if($nodes->length === 0) {
> > +            yield from $this->chatClient->postMessage(sprintf(
> > +                ":%s Did you know? That `%s...` dosen't exist in the world! Cuz' GOOGLE can't find it :P",
> > +                $message->getOrigin(),
> > +                substr(implode(" ", $message->getParameters()), 0, 60)
> > +            ));
> > +            return;
> > +        }
> > +        $length = min(3, $nodes->length);
> > +        $toPostMessage = "";
> > +        for($i = 0; $i < $length; $i++) {
> > +            start_of_loop:
>
> Label should be removed, since the goto is gone now.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/Room-11/Jeeves/pull/39/files/d13923ae5232bf3972349858a63ca13502fcd053#r60831861>
>